java-checks/src/main/java/org/sonar/java/checks/BoxedBooleanExpressionsCheck.java

```diff
@@ -127,17 +135,61 @@ private boolean isSafeBooleanExpression(ExpressionTree tree) {
   private static boolean isFirstUsageANullCheck(ExpressionTree boxedBoolean) {
     if (boxedBoolean.is(Kind.IDENTIFIER)) {
       IdentifierTree identifier = (IdentifierTree) boxedBoolean;
+      // Usages are not guaranteed to be ordered
       List<IdentifierTree> usages = identifier.symbol().usages();
-      if (!usages.isEmpty()) {
-        Tree parent = usages.get(0).parent();
-        if (parent.is(Kind.NOT_EQUAL_TO, Kind.EQUAL_TO) && isNullCheck((ExpressionTree) parent)) {
-          return true;
-        }
+      Tree firstUsage = usages.get(0).parent();
+      // Test if the first usage in our list is a null check
+      if (firstUsage.is(Kind.EQUAL_TO, Kind.NOT_EQUAL_TO) && isNullCheck((ExpressionTree) firstUsage)) {
+        return true;
+      }
+      // Return false if the only usage is not a null check
+      if (usages.size() == 1) {
+        return false;
+      }
+      // Fetch the first null check in the usages list
+      Optional<ExpressionTree> firstNullCheck = getFirstNullCheck(identifier);
+      if (!firstNullCheck.isPresent()) {
+        return false;
       }
+      // Test if the first null check and the first usage are part of the same higher if structure
+      Optional<IfStatementTree> ifStatementWithNullCheck = getParentConditionalBranch(firstNullCheck.get());
+      Optional<IfStatementTree> ifStatementWithFirstUsage = getParentConditionalBranch((ExpressionTree) firstUsage);
+      return ifStatementWithNullCheck.equals(ifStatementWithFirstUsage);
     }
     return false;
   }
 
+  private static Optional<ExpressionTree> getFirstNullCheck(IdentifierTree identifier) {
+    if (firstNullCheckCache.containsKey(identifier)) {
```

Comment:
If you look at the coverage report, you can notice that this condition will never be true.
It's because you're using 'IdentifierTree' as a key.

In case you want to cache the result for the variable, that might appear several times in the code, you need to use Symbol as a Key of map.

On the other hand, do you feel this situation is likely enough, so we need to cache? I don't know, so I am ok with and without caching.